os/kernel/binary_manager/binary_manager_data.c

```diff
@@ -206,35 +191,162 @@ binmgr_uinfo_t *binary_manager_get_udata(uint32_t bin_idx)
+	user_binary_header_t user_header_data;
+	char devpath[BINARY_PATH_LEN];
+
+	is_found = false;
+
+	bp_data = binary_manager_get_bpdata();
+	/* Update user binary data based on bootparam */
+	for (bp_app_idx = 0; bp_app_idx < bp_data->app_count; bp_app_idx++) {
+		bin_idx = binary_manager_get_index_with_name(bp_data->app_data[bp_app_idx].name);
+		if (bin_idx < 0) {
+			bmdbg("Failed to find matched binary %s in binary table \n", bp_data->app_data[bp_app_idx].name);
+			continue;
+		}
+		BIN_BPIDX(bin_idx) = bp_app_idx;
+		part_idx = bp_data->app_data[bp_app_idx].useidx;
+		snprintf(devpath, BINARY_PATH_LEN, BINMGR_DEVNAME_FMT, BIN_PARTNUM(bin_idx, part_idx));
+		if (bin_idx == BM_CMNLIB_IDX) {
+			ret = binary_manager_read_header(BINARY_COMMON, devpath, (void *)&common_header_data, false);			
+			version = common_header_data.version;
+		} else {
+			ret = binary_manager_read_header(BINARY_USERAPP, devpath, (void *)&user_header_data, false);			
+			version = user_header_data.bin_ver;
+		}
+		if (ret == OK) {
+			/* Return true it there is at least one valid binary */
+			BIN_USEIDX(bin_idx) = part_idx;
+			BIN_COUNT(bin_idx) = 1;
+			BIN_VER(bin_idx, part_idx) = version;			
+			bmdbg("[%d] part index %d, version %d\n", bin_idx, part_idx, version);
+			if (bin_idx != BM_CMNLIB_IDX) {
+				BIN_LOAD_PRIORITY(bin_idx, part_idx) = user_header_data.loading_priority;
+			}
+			/* Return true it there is at least one valid binary */
+			is_found = true;
+		} else {
+			bmdbg("Failed to find valid binary, %s\n", BIN_NAME(bin_idx));
+		}
+	}
+
+	return is_found;
+}
+
+/********************************************************************************************
+ * Name: binary_manager_check_user_update
+ *
+ * Description:
+ *	 This function checks that new user binary exists on inactive partition
+ *	and verifies the update is needed by comparing running version with new version.
+ *
+ *******************************************************************************************/
+int binary_manager_check_user_update(int bin_idx, bool *need_update)
+{
+	int ret;
+	int part_idx;
+	uint32_t version;
+	uint32_t running_ver;
+	char devpath[BINARY_PATH_LEN];
+	common_binary_header_t common_header_data;
+	user_binary_header_t user_header_data;
+
+	if (bin_idx < 0 || bin_idx > binary_manager_get_ucount()) {
+		bmdbg("Invalid bin idx %d\n", bin_idx);
+		return BINMGR_INVALID_PARAM;
+	}
+
+	*need_update = false;
+
+	running_ver = BIN_VER(bin_idx, BIN_USEIDX(bin_idx));
+	part_idx = BIN_USEIDX(bin_idx) ^ 1;
+
+	snprintf(devpath, BINARY_PATH_LEN, BINMGR_DEVNAME_FMT, BIN_PARTNUM(bin_idx, part_idx));
+	bmdbg("Devpath %s\n", devpath);
+	if (bin_idx == BM_CMNLIB_IDX) {
+		ret = binary_manager_read_header(BINARY_COMMON, devpath, (void *)&common_header_data, true);
+		version = common_header_data.version;
+	} else {
+		ret = binary_manager_read_header(BINARY_USERAPP, devpath, (void *)&user_header_data, true);
+		version = user_header_data.bin_ver;
+	}
+	if (ret == OK) {		
```

Comment:
Even if `ret` is not OK, but `binary_manager_check_user_update` can return `BINMGR_OK`.
Is it okay?